contracts/perun-vchannel-typescript/src/entry.rs

```diff
@@ -0,0 +1,1633 @@
+
+    // The channel constants do not change during the lifetime of a channel. They are located in the
+    // args field of the pcts.
+
+    let channel_constants =
+        ChannelConstants::from_slice(&args).expect("unable to parse args as ChannelConstants");
+    debug!("parsing channel constants passed");
+
+    debug!("channel_constants: {:?}", channel_constants);
+
+    // Verify that the channel parameters are compatible with the currently supported
+    // features of perun channels.
+    verify_vchannel_params_compatibility(&channel_constants.params())?;
+    debug!("verify_channel_params_compatibility passed");
+
+    // Next, we determine whether the transaction starts, progresses or closes the channel and fetch
+    // the respective old and/or new channel status.
+
+    let channel_action = get_vchannel_action()?;
+    debug!("get_channel_action passed");
+
+    match channel_action {
+        VChannelAction::Start {
+            new_vc_status,
+            old_lc_status,
+            new_lc_status,
+        } => {
+            debug!("Start action detected");
+            check_valid_vc_start(&new_vc_status, &channel_constants)
+        }
+        VChannelAction::Progress {
+            old_status,
+            new_status,
+        } => {
+            let channel_witness = load_witness()?;
+            debug!("load_witness passed");
+            check_valid_vc_progress(
+                &old_status,
+                &new_status,
+                &channel_witness,
+                &channel_constants,
+            )
+        }
+        VChannelAction::Close { old_status } => {
+            debug!("Close action detected");
+            let channel_witness = load_witness()?;
+            debug!("load_witness passed");
+            check_valid_vc_close(&old_status, &channel_witness, &channel_constants)
+        }
+    }
+}
+
+pub fn check_valid_vc_start(
+    new_status: &VirtualChannelStatus,
+    channel_constants: &ChannelConstants,
+) -> Result<(), Error> {
+    const FUNDER_INDEX: usize = 0;
+
+    debug!("check_valid_start");
+
+    // Upon start of a channel, the channel constants are stored in the args field of the pcts output.
+    // We uniquely identify a channel through the combination of the channel id (hash of ChannelParameters,
+    // which is part of the ChannelConstants) and the "thread token".
+    // The thread token contains an OutPoint and the channel type script verifies, that that outpoint is
+    // consumed in the inputs of the transaction that starts the channel.
+    // This means: Once a (pcts-hash, channel-id, thread-token) tuple appears once on chain and is recognized
+    // as the on-chain representation of this channel by all peers, no other "copy" or "fake" of that channel
+    // can be created on chain, as an OutPoint can only be consumed once.
+
+    // here, we verify that the OutPoint in the thread token is actually consumed.
+    verify_thread_token_integrity(&channel_constants.thread_token())?;
```

Comment:
Virtual channel cells should not have a thread token.
They gain their uniqueness through merging and the uniqueness of the parent cells.